### PR TITLE
Adding support of MAP_TOP_N_KEYS to Presto

### PR DESCRIPTION
**Pull request by jainavi17**
Add an UDF for getting top N keys of a map, which returns an array of the top N keys of map. Optionally a lambda comparator can also be passed to perform a custom comparison of the values. Returns all the value if the value N is greater than or equal to size of the map. N must be >= 0. For N = 0, the function returns empty array.

map_top_n_keys(map(K, V), N) -> array(K)

- For N > 0, returns the top N keys of a map. If N happens to larger than size of the map, it returns all the keys of the map in descending order
- For N = 0, return empty array
- For N < 0, function throws an error.

map_top_n_keys(map(K, V), N, function(K, K, int)) -> array(K)
Return the top N keys of the map sorted using the lambda comparator.

Test plan - (Please fill in how you tested your changes)
Added unit tests.
Build successfully using the following terminal command
- `./mvnw clean install -Dtest=TestMapTopNKeysFunction -fn -pl presto-main`
- `./mvnw clean install -Dtest=TestMapTopNKeysComparatorFunction -fn -pl presto-main`

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==

General Changes
* Add support for MAP_TOP_N_KEYS
* Returns an array of the top N keys of map. Optionally a lambda comparator can also be passed to perform a custom comparison of the keys. Returns all the keys if the value N is greater than or equal to size of the map. For N < 0, the function returns keys in the reverse order. For N = 0, the function returns empty array.

.. function:: map_top_n_keys(map(K, V), N) -> array(K)
   Returns the top N keys of ``map``. If N must be >= 0. For N = 0, it returns empty ``array``.

.. function:: map_top_n_keys(map(K, V), N, function(K, K, int)) -> array(K)
   Returns the top N keys of ``map`` sorted using lambda comparator. N must be >= 0. For N = 0, it returns empty ``array``.

`SELECT MAP_TOP_N_KEYS(MAP(ARRAY[4, 5, 6], ARRAY[1, 2, 3]), 2)); -- 6, 5`
`SELECT MAP_TOP_N_KEYS(MAP(ARRAY['a', 'b', 'c', 'd'], ARRAY[1, 2, 3, NULL, NULL]), -3)); -- 'a', 'b', 'c'`
`SELECT MAP_TOP_N_KEYS(MAP(ARRAY[4, 5, 6], ARRAY[1, 2, 3]), 4, (x, y) -> IF(x < y, 1, IF(x = y, 0, -1))); -- 4, 5, 6`
```

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java [hunk 1]
```diff
@@ -37,6 +38,7 @@ public void testBasic()
     @Test
     public void testNonPositiveN()
     {
+        assertFunction("MAP_TOP_N_KEYS(MAP(ARRAY[100, 200, 300], ARRAY[4, 5, 6]), 0)", new ArrayType(INTEGER), asList());
         assertFunction("MAP_TOP_N_KEYS(MAP(ARRAY[1, 2, 3], ARRAY[4, 5, 6]), -1)", new ArrayType(INTEGER), asList(1));
```
[the remark this answers is not in the document]
Comment (jainavi17) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java:42]: Sure let me add that.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    @SqlInvokedScalarFunction(value = "map_top_n_keys", deterministic = true, calledOnNullInput = true)
+    @Description("Returns the top N keys of the given map in sorted using the provided lambda comparator.")
+    @TypeParameter("K")
+    @TypeParameter("V")
+    @SqlParameters({@SqlParameter(name = "input", type = "map(K, V)"), @SqlParameter(name = "n", type = "bigint"), @SqlParameter(name = "f", type = "function(K, K, int)")})
+    @SqlType("array<K>")
+    public static String mapTopNKeysComparator()
+    {
+        return "RETURN IF(n >= 0, slice(reverse(array_sort(map_keys(input), f)), 1, n), slice(array_sort(map_keys(input), f), 1, ABS(n)))";
+    }
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:47]: Also you don't need the ABS etc., slice works both ways - if you give negative index it goes reverse. so just slice(.. ,1, n) should do it.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    public void testEmpty()
+    {
+        assertFunction("MAP_TOP_N_KEYS(MAP(ARRAY[], ARRAY[]), 5)", new ArrayType(UNKNOWN), asList());
+    }
+
+    @Test
+    public void testNull()
+    {
+        assertFunction("MAP_TOP_N_KEYS(NULL, 1)", new ArrayType(UNKNOWN), null);
+    }
```
Comment (kaikalur) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java:56]: Ans also a bad comparator to make sure it gives error properly

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jainavi17) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java:42]: So just to get it right, this is only for keys. So I have added both empty keys and null map in the test cases. We cannot have nulls in the key and since we're not looking for values so not sure if its gonna create any difference.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java (hunk 2 above)
Comment (jainavi17) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:47]: I think in slice negative index still just set the pointer from reverse direction but the slice still gives you n elements starting from that element

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java (hunk 3 above)
Comment (jainavi17) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java:56]: OK let me add it to TestMapTopNKeysComparator

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java (hunk 2 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:47]: > I think in slice negative index still just set the pointer from reverse direction but the slice still gives you n elements starting from that element

yeah - in that case start with -N?

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:47]: So maybe just do:

SLICE(..., LEAST(1, -N), N)

Comment (jainavi17) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:47]: Hmm...I can do this but not sure if it will have any advantage over what I'm already doing. So using the above method for N > 0, I still need to reverse the array. For example

SELECT SLICE(ARRAY[1, 2, 3, 4], LEAST(1, -3), 3) -> output: [2, 3, 4 ], expected: [4, 3, 2]

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:47]: > Hmm...I can do this but not sure if it will have any advantage over what I'm already doing. So using the above method for N > 0, I still need to reverse the array. For example
> 
> SELECT SLICE(ARRAY[1, 2, 3, 4], LEAST(1, -3), 3) -> output: [2, 3, 4 ], expected: [4, 3, 2]

Is that the expected? I'm not sure. In fact, I'm beginning to think we disallow N < 0. Let them define a comparator that reverses the direction. That saves a lot of confusion. So I would do:

IF(N < 0, fail('N must be non-negative'), ...)

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:47]: Also let's update the documentation to say if they want bottom N, they need a comparator that reverses the direction

Comment (jainavi17) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:47]: We can disallow negative N but then every time if need bottom N keys even for simple key data types, people do need to pass the lambda. But yeah I am happy to throw error in case N <= 0.  Let me know

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:47]: That's fine. We need clean semantics in the language and sometimes that makes things hard for some use cases. We can always add MAP_BOTTOM_N_KEYS as a separate udf if needed - especially since we now have SQL UDFs supported better.

Comment (jainavi17) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:47]: Cool!

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    @Test
+    public void testEmpty()
+    {
+        assertFunction("MAP_TOP_N_KEYS(MAP(ARRAY[], ARRAY[]), 5)", new ArrayType(UNKNOWN), asList());
+    }
+
+    @Test
+    public void testNull()
+    {
+        assertFunction("MAP_TOP_N_KEYS(NULL, 1)", new ArrayType(UNKNOWN), null);
```
Comment (kaikalur) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java:55]: We also need tests where the values are null - to demonstrat NULLs first/last semantics. Add something like:

MAP_TOP_N_VALUES(MAP(ARRAY[1,2,3,4], ARRAY[1,2,null, null])) 

And see what you get.

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java (hunk 4 above)
Comment (kaikalur) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java:55]: Oops never mind. You are doing only keys!

**Review by kaikalur — APPROVED** [review]

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMapTopNKeysComparatorFunction.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        assertInvalidFunction("MAP_TOP_N_KEYS(MAP(ARRAY[4, 5, 6], ARRAY[1, 2, 3]), 0, (x, y) -> IF(x < y, 1, IF(x = y, 0, -1)))", StandardErrorCode.GENERIC_USER_ERROR, "n must be a positive integer");
+        assertInvalidFunction("MAP_TOP_N_KEYS(MAP(ARRAY[4, 5, 6], ARRAY[1, 2, 3]), -2, (x, y) -> IF(x > y, 1, IF(x = y, 0, -1)))", StandardErrorCode.GENERIC_USER_ERROR, "n must be a positive integer");
+        assertInvalidFunction("MAP_TOP_N_KEYS(MAP(ARRAY['x', 'y', 'z'], ARRAY[1, 2, 3]), -3, (x, y) -> IF(x > y, 1, IF(x = y, 0, -1)))", StandardErrorCode.GENERIC_USER_ERROR, "n must be a positive integer");
+    }
+
+    @Test
+    public void testEmpty()
+    {
+        assertFunction("MAP_TOP_N_KEYS(MAP(ARRAY[], ARRAY[]), 1, (x, y) -> IF(x < y, 1, IF(x = y, 0, -1)))", new ArrayType(UNKNOWN), asList());
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMapTopNKeysComparatorFunction.java:65]: use `emptyList()` or `ImmutableList.of()`

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+public class TestMapTopNKeysFunction
+        extends AbstractTestFunctions
+{
+    @Test
+    public void testBasic()
+    {
+        assertFunction("MAP_TOP_N_KEYS(MAP(ARRAY[1, 2, 3], ARRAY[4, 5, 6]), 2)", new ArrayType(INTEGER), asList(3, 2));
+        assertFunction("MAP_TOP_N_KEYS(MAP(ARRAY[-1, -2, -3], ARRAY[4, 5, 6]), 2)", new ArrayType(INTEGER), asList(-1, -2));
+        assertFunction("MAP_TOP_N_KEYS(MAP(ARRAY['ab', 'bc', 'cd'], ARRAY['x', 'y', 'z']), 1)", new ArrayType(createVarcharType(2)), asList("cd"));
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java:37]: use `singletonList(..)` or `ImmutableList.of(..)`

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        assertInvalidFunction("MAP_TOP_N_KEYS(MAP(ARRAY[100, 200, 300], ARRAY[4, 5, 6]), 0)", StandardErrorCode.GENERIC_USER_ERROR, "n must be a positive integer");
+        assertInvalidFunction("MAP_TOP_N_KEYS(MAP(ARRAY[1, 2, 3], ARRAY[4, 5, 6]), -1)", StandardErrorCode.GENERIC_USER_ERROR, "n must be a positive integer");
+        assertInvalidFunction("MAP_TOP_N_KEYS(MAP(ARRAY['a', 'b', 'c'], ARRAY[4, 5, 6]), -2)", StandardErrorCode.GENERIC_USER_ERROR, "n must be a positive integer");
+    }
+
+    @Test
+    public void testEmpty()
+    {
+        assertFunction("MAP_TOP_N_KEYS(MAP(ARRAY[], ARRAY[]), 5)", new ArrayType(UNKNOWN), asList());
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNKeysFunction.java:52]: use `emptyList()` or `ImmutableList.of()`. Same for all other places in the PR

**Review by highker — APPROVED** [review]

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMapTopNKeysComparatorFunction.java (hunk 5 above)
Comment (jainavi17) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMapTopNKeysComparatorFunction.java:65]: Cool! Doing it

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    @SqlInvokedScalarFunction(value = "map_top_n_keys", deterministic = true, calledOnNullInput = true)
+    @Description("Returns the top N keys of the given map in sorted using the provided lambda comparator.")
+    @TypeParameter("K")
+    @TypeParameter("V")
+    @SqlParameters({@SqlParameter(name = "input", type = "map(K, V)"), @SqlParameter(name = "n", type = "bigint"), @SqlParameter(name = "f", type = "function(K, K, int)")})
+    @SqlType("array<K>")
+    public static String mapTopNKeysComparator()
+    {
+        return "RETURN IF(n <= 0, fail('n must be a positive integer'), slice(reverse(array_sort(map_keys(input), f)), 1, n))";
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:46]: Actually, 0 is fine - should return empty array.

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java (hunk 8 above)
Comment (jainavi17) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:46]: Yes its fine but just thinking that what's the point of allowing 0 that means users don't want to return to anything which doesn't make sense. What do you think?

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:46]: Again - language semantics should be clean, precise and consistent. We should allow 0 like all others like SLICE

Comment (jainavi17) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:46]: Hmm..cool let me add it back quickly

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    @SqlInvokedScalarFunction(value = "map_top_n_keys", deterministic = true, calledOnNullInput = true)
+    @Description("Returns the top N keys of the given map in sorted using the provided lambda comparator.")
+    @TypeParameter("K")
+    @TypeParameter("V")
+    @SqlParameters({@SqlParameter(name = "input", type = "map(K, V)"), @SqlParameter(name = "n", type = "bigint"), @SqlParameter(name = "f", type = "function(K, K, int)")})
+    @SqlType("array<K>")
+    public static String mapTopNKeysComparator()
+    {
+        return "RETURN IF(n < 0, fail('n must be a positive integer'), slice(reverse(array_sort(map_keys(input), f)), 1, n))";
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:46]: message: n must be a >= 0 - so more explict

**Review by kaikalur — APPROVED** [review]

**Comment by jainavi17** [comment]
Can we merge this?
